src/Service/VendorService/OverDrive/Api/Client.php

```diff
@@ -0,0 +1,316 @@
+use Psr\Cache\InvalidArgumentException;
+use Symfony\Component\Cache\Adapter\AdapterInterface;
+
+/**
+ * Class Client.
+ */
+class Client
+{
+    private const MAX_RETRIES = 5;
+
+    private const OAUTH_URL = 'https://oauth.overdrive.com/token';
+    private const USER_AGENT = 'cover.dandigbib.org';
+
+    /** @var string */
+    private $libraryAccountEndpoint;
+
+    /** @var string */
+    private $clientId;
+
+    /** @var string */
+    private $clientSecret;
+
+    /** @var AdapterInterface */
+    private $cache;
+
+    /** @var ClientInterface */
+    private $httpClient;
+
+    /** @var string */
+    private $authorization;
+
+    /** @var string */
+    private $productsEndpoint;
+
+    /**
+     * Client constructor.
+     *
+     * @param AdapterInterface $cache
+     * @param ClientInterface $httpClient
+     */
+    public function __construct(AdapterInterface $cache, ClientInterface $httpClient)
+    {
+        $this->cache = $cache;
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * Set OverDrive authentication credentials.
+     *
+     * @see https://developer.overdrive.com/apis/client-auth
+     * @see https://developer.overdrive.com/apis/library-account
+     *
+     * @param string $libraryAccountEndpoint
+     *   The custom endpoint for the account used
+     * @param string $clientId
+     *   Client id
+     * @param string $clientSecret
+     *   Client secret
+     */
+    public function setCredentials(string $libraryAccountEndpoint, string $clientId, string $clientSecret): void
+    {
+        $this->libraryAccountEndpoint = $libraryAccountEndpoint;
+        $this->clientId = $clientId;
+        $this->clientSecret = $clientSecret;
+    }
+
+    /**
+     * Get cover url from OverDrive crossRefId.
+     *
+     * @param string $crossRefId
+     *   The OverDrive 'crossRefId'
+     *
+     * @return string|null
+     *   The cover url or null
+     *
+     * @throws GuzzleException
+     * @throws AuthException
+     * @throws InvalidArgumentException
+     */
+    public function getCoverUrl(string $crossRefId): ?string
+    {
+        $requests = 0;
+
+        // The OverDrive api fails at random but still gives a 200 response.
+        // Only indication of failure is that the 'products' key is missing
+        // from the response body.
+        // If the products key is missing we retry the request 'MAX_RETRIES' times
+        do {
+            $response = $this->httpClient->request('GET', $this->getProductsEndpoint(), [
+                'headers' => $this->getHeaders(),
+                'query' => [
+                    'crossRefId' => $crossRefId,
+                ],
+            ]);
+
+            $content = $response->getBody()->getContents();
+            $json = json_decode($content, false);
+
+            // Check for the product and images keys.
+            $product = isset($json->products) && is_array($json->products) ? array_shift($json->products) : null;
```

Comment:
If the response is not 200 or json. The variable `$json` would be `false` and accessing a property on false will throw an exception.

Reply:
Added `JSON_THROW_ON_ERROR` and `catch` clause